### PR TITLE
Fix note collisions for 3.0

### PR DESCRIPTION
**Pull request by mirabilos**
Same as #3514 for `master`.

Please test first, I do not currently have a development environment for `master` (will do after 2.2.0 is released).

### CODE REVIEW

**Comment by mirabilos** [comment]
Merged #3551 and squashed all commits into one since @lasconic seems to prefer that

**Comment by lasconic** [comment]
Let's used the approach in https://github.com/musescore/MuseScore/pull/3560 for now.

**Comment by mirabilos** [comment]
• Rebased on current master (as of today)
• used the approach in #3560 over #3579 for now, but in a second commit, easily `git revert`able later once we *do* provide channel assignment

**Comment by lasconic** [comment]
So we need https://github.com/musescore/MuseScore/pull/3579 and https://github.com/musescore/MuseScore/pull/3588 in this PR to be on par with 2.2.1 / 2.3

**Comment by mirabilos** [comment]
@lasconic rebased onto latest master and merged the two PRs into the respective correct commits (first commit to stay, second commit to be `git revert`ed later when we switch back from restriking)

**Comment by mirabilos** [comment]
Rebased against latest master, keeping the split into two commits (so the second one can later easily be reverted), check second commit against https://github.com/musescore/MuseScore/pull/3797 and remove one useless/unused variable.

File: mtest/libmscore/midi/tst_midi.cpp [hunk 1]
```diff
@@ -378,6 +378,7 @@ void TestMidi::events()
       QTextStream out(&filehandler);
 
       for (auto iter = events.begin(); iter!= events.end(); ++iter){
+            if (iter->second.discard()) continue;
```
Comment (anatoly-os) [inline mtest/libmscore/midi/tst_midi.cpp]: When rebasing, please put `continue` at the next line.

File: synthesizer/event.cpp [hunk 2]
```diff
@@ -372,5 +372,51 @@ void EventList::insert(const Event& e)
...
+      /* track info for each of the 128 possible MIDI notes */
+      struct channelInfo {
+            /* which event the first ME_NOTEON came from */
+            NPlayEvent *event[128];
+            /* how often is the note on right now? */
+            unsigned short nowPlaying[128];
+            };
+
+      /* track info for each channel (on the heap, 0-initialised) */
+      struct channelInfo *info = (struct channelInfo *)calloc(_highestChannel + 1, sizeof(struct channelInfo));
```
Comment (anatoly-os) [inline synthesizer/event.cpp]: Can we avoid c-like allocations using C++11 style instead like using `new()` operator or even `std::unique_ptr`? The last is known to work without side effects on memory allocation.

File: libmscore/rendermidi.cpp [hunk 3]
```diff
@@ -1780,22 +1791,29 @@ void Score::renderMetronome(EventMap* events, Measure* m, int tickOffset)
...
-      updateRepeatList(MScore::playRepeats);
-      _foundPlayPosAfterRepeats = false;
+      updateRepeatList(expandRepeats);
       masterScore()->updateChannel();
       updateVelo();
 
       // create note & other events
       for (Staff* part : _staves)
             renderStaff(events, part);
+      events->fixupMIDI();
```
Comment (anatoly-os) [inline libmscore/rendermidi.cpp]: Did you track performance impact of this change? Afaics it adds some computations to on fly generation of the audio.

**Comment by anatoly-os** [comment]
@mirabilos I left few comments on this PR. Sorry, I absolutely forgot we need to correctly process such kind of notes collisions. Let's finish this PR so that it could be merged.

**Comment by mirabilos** [comment]
anatoly-os dixit:

>@mirabilos I left few comments on this PR. Sorry, I absolutely forgot
>@we need to correctly process such kind of notes collisions. Let's

Indeed.

>@finish this PR so that it could be merged.

Can we please first get this merged so we have the (exact) same
functionality from 2.x ported to master, and only then change
the code? It’s work enough keeping track of the 2.x changes,
rebasing them against master, etc. all the time, and I’m glad I
managed to dig out everything involved.

Additionally, these two commits are comprised of code from me
and lasconic. I don’t want to add more complications to that,
and the code clearly was good enough to have been shipped in
actual releases, so it should be good to merge as-is.

While I agree with…
>When rebasing, please put `continue` at the next line.
and almost certainly wouldn’t have written that myself (so
this is probably from one of lasconic’s commits), there is
such a thing called “separation of concerns”. This PR very
much concerns itself with bringing master up-to-date with
2.x and really ought to have been dealt with within DAYS,
at most, of the related commits in 2.x.

>Can we avoid c-like allocations using C++11 style instead like using
>`new()` operator or even `std::unique_ptr`? The last is known to work
>without side effects on memory allocation.

This is one of the things I’d prefer to leave to people
with more C++ experience than myself. Another reason to
have fixups later. I’m not capable enough to fulfil this,
if it were formulated as requirement, and so master would
lose bugfixes from 2.x.

>Did you track performance impact of this change? Afaics it adds some
>computations to on fly generation of the audio.

Yes, we had some discussion on that, with multiple people
involved, but I don’t recall when or on what medium exactly.
Probably IRC. It’s been some months or so since then, after all.

I do recall that the performance impact was less than we’d
have expected, and acceptable to all involved.

Also, again, this PR is *exactly* about making master catch
up with the bugfixes that have been shipping for *several*
releases by now, and as such is more than overdue, especially
in its current form. (The squash was requested by lasconic,
and the split into two commits is needed for upcoming changes.)

File: mtest/libmscore/midi/tst_midi.cpp (hunk 1 above)
Comment (mirabilos) [reply mtest/libmscore/midi/tst_midi.cpp]: Done in rebasing.

**Comment by anatoly-os** [comment]
AppVeyor was not even triggered. Could you push some small changes?

**Comment by mirabilos** [comment]
@anatoly-os  is there anything holding off merging this? I had to rebase it *again*…

**Comment by anatoly-os** [comment]
My last question is not addressed yet. My last concern is about possible performance drop in playback time due to introducing fixupMIDI() method.

**Comment by mirabilos** [comment]
We analysed this before the patch was accepted in 2.x and all involved people agreed that the performance impact was neglegible.

**Pull request merged**